README.md
[![Build Status](https://travis-ci.org/n3okill/enfsfind.svg)](https://travis-ci.org/n3okill/enfsfind)
[![Build status](https://ci.appveyor.com/api/projects/status/w8567ywc795x06po?svg=true)](https://ci.appveyor.com/project/n3okill/enfsfind)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/125b0fd16c4e4ad390acd52b216586e1)](https://www.codacy.com/app/n3okill/enfsfind)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfsfind.png)](https://nodei.co/npm/enfsfind/)

enfsfind
=========
Module that add find functionality to node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)


Description
-----------
This module will add a method that allows the obtaining of a
list of items in the file system under one directory and sub-directories
filtering the result.

- This module will add following methods to node fs module:
  * find
  * findSync
  
Usage
-----
`enfsfind`

```js
    const enfsfind = require("enfsfind");
```

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [find](#find)
- [findSync](#findsync)


### find
  - **find(path, [options], callback)**

> Obtain the list of items under a directory and sub-directories asynchronously.
Each item will be an object containing: {path: pathToItem, stat: itemStat}

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfslist](https://www.npmjs.com/package/enfslist))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * filter (Function or RegExp): if defined will filter the items in the file system

No Filter:

```js
    enfsfind.find("/home", function(err, listOfItems){
        listOfItems.forEach(function(item){
            //do something
        });
    });
```

Filter (Function):

```js
    function filterFn(path, stat){
        return stat.isFile() && path.indexOf("mp3") !== -1;
    }
    enfsfind.find("/home", {filter: filterFn},function(err, listOfItems){
        listOfItems.forEach(function(item){
            //do something
        });
    });
```

Filter (RegExp):

```js
    enfsfind.find("/home", /\.mp3$/,function(err, listOfItems){
        listOfItems.forEach(function(item){
            //do something
        });
    });
```


### findSync
  - **findSync(path, [options])**

> Obtain the list of items under a directory and sub-directories synchronously
Each item will be an object containing: {path: pathToItem, stat: itemStat}

[options]:
  * fs (Object): an alternative fs module to use (default will be [enfslist](https://www.npmjs.com/package/enfslist))
  * dereference (Boolean): if true will dereference symlinks listing the items to where it points (default: false)
  * filter (Function or RegExp): if defined will filter the items in the file system

No Filter:

```js
    const listOfItems = enfsfind.findSync("/home");
    listOfItems.forEach(function(item){
        //do something
    });
```

Filter (Function):

```js
    function filterFn(path, stat){
        return stat.isFile() && path.indexOf("mp3") !== -1;
    }
    const listOfItems = enfsfind.findSync("/home", {filter: filterFn});
    listOfItems.forEach(function(item){
        //do something
    });
```

Filter (RegExp):

```js
    const listOfItems = enfsfind.findSync("/home", /\.mp3$/);
    listOfItems.forEach(function(item){
        //do something
    });
```

License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-03-31
==================
  * code improvement
  * dependencies update
  * major version

0.1.0 / 2016-04-06
==================
  * es6 update
  * change in filter method arguments

0.0.1 / 2016-02-20
==================

  * first release


test/findSync.js
/**
 * @project enfsfind
 * @filename findSync.js
 * @description sync methods for finding items inside directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */
"use strict";

const nodePath = require("path");
const find = require("../");

describe("enfsfind sync", function() {
    const tmpPath = nodePath.join(__dirname, "..", "lib");
    it("should list files", function() {
        find.findSync(tmpPath, "").length.should.be.equal(4);
    });
    it("should list and filter files with regex", function() {
        find.findSync(tmpPath, /async/gi).length.should.be.equal(1);
    });
    it("should list and filter files with function", function() {
        find.findSync(tmpPath, function(path) {
            return /sync/gi.test(path);
        }).length.should.be.equal(2);
    });
    it("should list and filter files with regex", function() {
        find.findSync(tmpPath, /lib/).length.should.be.equal(4);
    });
});


test/findAsync.js
/**
 * @project enfsfind
 * @filename findAsync.js
 * @description sync methods for finding items inside directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

"use strict";

const nodePath = require("path");
const find = require("../");

describe("enfsfind async", function() {
    const tmpPath = nodePath.join(__dirname, "..", "lib");
    it("should list files", function(done) {
        find.find(tmpPath, function(err, list) {
            (err === null).should.be.equal(true);
            list.length.should.be.equal(4);
            done();
        });
    });
    it("should list and filter files with regex", function(done) {
        find.find(tmpPath, /async/gi, function(err, list) {
            (err === null).should.be.equal(true);
            list.length.should.be.equal(1);
            done();
        });
    });
    it("should list and filter files with function", function(done) {
        find.find(tmpPath, function(path) {
            return /sync/gi.test(path);
        }, function(err, list) {
            (err === null).should.be.equal(true);
            list.length.should.be.equal(2);
            done();
        });
    });
    it("should list and filter files with regex", function(done) {
        find.find(tmpPath, /lib/, function(err, list) {
            (err === null).should.be.equal(true);
            list.length.should.be.equal(4);
            done();
        });
    });
});


lib/findSync.js
/**
 * @project enfsfind
 * @filename findSync.js
 * @description sync methods for finding items inside directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */


"use strict";

const nodeAssert = require("assert");
const enFs = require("enfspatch");
const enfslist = require("enfslist");

const kindOf = (arg) => arg === null ? "null" : typeof arg === "undefined" ? "undefined" : /^\[object (.*)\]$/.exec(Object.prototype.toString.call(arg))[1];
const isFunction = (arg) => "Function".toLowerCase() === kindOf(arg).toLowerCase();
const isRegExp = (arg) => "RegExp".toLowerCase() === kindOf(arg).toLowerCase();


/**
 * Find all the items inside a directory and sub-directories based on a provided filter
 * @param {string} path - the path to the directory
 * @param {object} opt - various options for list module
 *              {bool} options.dereference - defines the type of stat that will be used in files
 *              {object} options.fs - the fs module to be used
 *              {function|RegExp} options.filter - function or RegExp used to narrow results
 *              this function receives an object of type {path: itemPath, stat: itemStatObject}
 * @return {Error|Array}
 */

function find(path, opt) {
    let options, result;
    nodeAssert(path, "path is required");

    if (opt) {
        if (isFunction(opt) || isRegExp(opt)) {
            options = {filter: opt};
        } else {
            options = opt;
        }
    }
    options = options || {};
    options.fs = options.fs || enFs;
    options.dereference = options.dereference === true;
    result = enfslist.listSync(path, options);
    if (options.filter) {
        if (isRegExp(options.filter)) {
            result = result.filter((item) => options.filter.test(item.path));
        } else if (isFunction(options.filter)) {
            result = result.filter((item) => options.filter(item.path, item.stats));
        }
    }
    return result;
}


module.exports = find;



lib/findAsync.js
/**
 * @project enfsfind
 * @filename findAsync.js
 * @description async methods for finding items inside directories
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.2
 */


"use strict";


const nodeAssert = require("assert");
const enFs = require("enfspatch");
const enfslist = require("enfslist");


const kindOf = (arg) => arg === null ? "null" : typeof arg === "undefined" ? "undefined" : /^\[object (.*)\]$/.exec(Object.prototype.toString.call(arg))[1];
const isFunction = (arg) => "Function".toLowerCase() === kindOf(arg).toLowerCase();
const isRegExp = (arg) => "RegExp".toLowerCase() === kindOf(arg).toLowerCase();


/**
 * Find all the items inside a directory and sub-directories based on a provided filter
 * @param {string} path - the path to the directory
 * @param {object} opt - various options for list module
 *              {bool} options.dereference - defines the type of stat that will be used in files
 *              {object} options.fs - the fs module to be used
 *              {function|RegExp} options.filter - function or RegExp used to narrow results
 *              this function receives an object of type {path: itemPath, stat: itemStatObject}
 * @param {function} callback - the callback function that will be called after the list is done
 * @return {Error|Array}
 */

function find(path, opt, callback) {
    let options;
    nodeAssert(path, "path is required");
    if (isFunction(opt) && !callback) {
        callback = opt;
        opt = null;
    }

    if (opt) {
        if (isFunction(opt) || isRegExp(opt)) {
            options = {filter: opt};
        } else {
            options = opt;
        }
    }
    options = options || {};
    nodeAssert(typeof callback === "function", "callback must be a function");

    options.fs = options.fs || enFs;
    options.dereference = options.dereference === true;

    enfslist.list(path, options, (err, list) => {
        if (err) {
            return callback(err);
        }
        if (options.filter) {
            if (isRegExp(options.filter)) {
                list = list.filter((item) => options.filter.test(item.path));
            } else if (isFunction(options.filter)) {
                list = list.filter((item) => options.filter(item.path, item.stats));
            }
        }
        callback(null, list);
    });
}


module.exports = find;

